garak/_plugins.py

```diff
@@ -58,40 +62,105 @@ def _extract_modules_klasses(base_klass):
+                    )
+                    if os.path.exists(file_path):
+                        validated_plugin_filenames.add(file_path)
+                        mod_time = datetime.fromtimestamp(
+                            os.path.getmtime(file_path), tz=timezone.utc
+                        )
+                    else:
+                        # file not found force cache rebuild
+                        mod_time = datetime.now(timezone.utc)
+                    if mod_time > user_time:
+                        # rebuild all for now, this can be more made more selective later
+                        return False
+
+            # if all known are up-to-date check filesystem for missing
+            found_filenames = set()
+            plugin_path = _config.transient.package_dir / category
+            for module_filename in sorted(os.listdir(plugin_path)):
+                if not module_filename.endswith(".py"):
+                    continue
+                if module_filename.startswith("__"):
+                    continue
+                found_filenames.add(plugin_path / module_filename)
+            if found_filenames != validated_plugin_filenames:
+                return False
+
+        return True  # all checks passed the cache is valid
 
     def _build_plugin_cache(self):
         """build a plugin cache file to improve access times
 
         This method writes only to the user's cache (currently the same as the system cache)
         TODO: Enhance location of user cache to enable support for in development plugins.
         """
-        local_cache = {}
-
-        for plugin_type in PLUGIN_TYPES:
-            plugin_dict = {}
-            for plugin in self._enumerate_plugin_klasses(plugin_type):
-                plugin_name = ".".join([plugin.__module__, plugin.__name__]).replace(
-                    "garak.", ""
-                )
-                plugin_dict[plugin_name] = PluginCache.plugin_info(plugin)
-
-            sorted_keys = sorted(list(plugin_dict.keys()))
-            local_cache[plugin_type] = {i: plugin_dict[i] for i in sorted_keys}
-
-        with open(
-            self._user_plugin_cache_filename, "w", encoding="utf-8"
-        ) as cache_file:
-            json.dump(local_cache, cache_file, cls=PluginEncoder, indent=2)
+        logging.debug("rebuilding plugin cache")
+        with PluginCache._mutex:
+            local_cache = {}
+
+            for plugin_type in PLUGIN_TYPES:
+                plugin_dict = {}
+                for plugin in self._enumerate_plugin_klasses(plugin_type):
+                    plugin_name = ".".join(
+                        [plugin.__module__, plugin.__name__]
+                    ).replace("garak.", "")
+                    plugin_dict[plugin_name] = PluginCache.plugin_info(plugin)
+
+                sorted_keys = sorted(list(plugin_dict.keys()))
+                local_cache[plugin_type] = {i: plugin_dict[i] for i in sorted_keys}
```

Comment:
is this sorting redundant, seeing as the data goes into a set?

Reply:
The sorting reduces churn in the `package` file when we commit updates. In testing of the original addition of the cache I found the order of plugins is not consistent unless sorted.

Reply:
ah, ok. this might be relying on non-guaranteed behavior in `dict` but i guess we can deal with that if it rears its head.